File: kotlin/abc087_b.kt
// ABC087B - Coins

fun main() {
  val A = readLine()!!.toInt()
  val B = readLine()!!.toInt()
  val C = readLine()!!.toInt()
  val X = readLine()!!.toInt()

  var cnt = 0
  repeat(A + 1) repeatA@ { a ->
    repeat(B + 1) repeatB@ { b ->
      val c50 = X - (a * 500 + b * 100)
      if (c50 < 0) return@repeatA
      if (c50 % 50 != 0 || c50 / 50 > C) return@repeatB
      cnt++
    }
  }
  println(cnt)
}

File: ruby/abc087_b.rb
# ABC087B - Coins

A = gets.to_i
B = gets.to_i
C = gets.to_i
X = gets.to_i

cnt = 0
(A + 1).times do |a|
  break if X < 500 * a
  (B + 1).times do |b|
    c50 = X - (500 * a + 100 * b)
    break if c50 < 0
    cnt += 1 if c50 <= C * 50 && c50 % 50 == 0
  end
end

puts cnt
